File: Data Structures & Algorithms/eating-bananas/submission-2.java
class Solution {

    public int func(int[]arr, int boundry){
        int total=0;
        for(int i=0;i<arr.length;i++){
            total+=Math.ceil((double)arr[i]/(double)boundry);
        }
        return total;
    }

    public int minEatingSpeed(int[] piles, int h) {
        Arrays.sort(piles);
        int left = 1,right=0;
        for(int pile : piles){
            right = Math.max(right,pile);
        }
        int ans = right;
        while(left<=right){
            int mid = left+(right-left)/2;
            int totalhours = func(piles,mid);
            if(totalhours<=h){
                ans = mid;
                right=mid-1;
            }else{
                left = mid+1;
            }
        }
        return ans;
    }
}

File: Data Structures & Algorithms/climbing-stairs/submission-2.cpp
class Solution {
public:

    int func(int n,vector<int> &arr){
        if(n<=1) return arr[n];
        if(arr[n]!=0) return arr[n];
        return arr[n] = func(n-1,arr)+func(n-2,arr);
    }

    int climbStairs(int n) {
        vector<int> arr(n+1,0);
        arr[0]=1;
        arr[1]=1;
        return func(n,arr);
    }
};
